crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -7113,13 +7152,56 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         }
     }
 
-    fn infer_and_check_argument_types(
+    fn infer_and_try_call_dunder<'a, I, F>(
         &mut self,
-        ast_arguments: &ast::Arguments,
+        db: &'db dyn Db,
+        place: Place<'db>,
+        ast_arguments: I,
+        argument_types: &mut CallArguments<'_, 'db>,
+        infer_argument_ty: F,
+        call_expression_tcx: TypeContext<'db>,
+    ) -> Result<Bindings<'db>, CallDunderError<'db>>
+    where
+        I: IntoIterator<Item = ArgOrKeyword<'a>> + Clone,
+        F: FnMut(&mut Self, usize, &ast::Expr, TypeContext<'db>) -> Type<'db>,
+    {
+        match place {
+            Place::Defined(dunder_callable, _, boundness, _) => {
+                let mut bindings = dunder_callable
+                    .bindings(db)
+                    .match_parameters(db, argument_types);
+
+                if let Err(call_error) = self.infer_and_check_argument_types(
+                    ast_arguments,
+                    argument_types,
+                    infer_argument_ty,
+                    &mut bindings,
+                    call_expression_tcx,
+                ) {
+                    return Err(CallDunderError::CallError(call_error, Box::new(bindings)));
+                }
+
+                if boundness == Definedness::PossiblyUndefined {
+                    return Err(CallDunderError::PossiblyUnbound(Box::new(bindings)));
+                }
+                Ok(bindings)
+            }
+            Place::Undefined => Err(CallDunderError::MethodNotAvailable),
+        }
+    }
+
+    fn infer_and_check_argument_types<'a, I, F>(
+        &mut self,
+        ast_arguments: I,
         argument_types: &mut CallArguments<'_, 'db>,
+        mut infer_argument_ty: F,
```

Comment:
Can we avoid monomorphizing the entire function here by e.g. making `infer_argument_ty` an enum? The iterator seems a bit more tricky unless we implement the iterator manually within `ruff_python_ast`